File: tools/agentstream/checklist/tracker_migration_preview.go
// Read-only Tracker migration preview comparing definition slug inventory to existing Tracker.
// [IMPL-TIED_CHECKLIST_GATE_ENFORCEMENT] [ARCH-TIED_CHECKLIST_GATE_ENFORCEMENT] [REQ-TIED_CHECKLIST_GATE_ENFORCEMENT]
// How: PREVIEW_TRACKER_MIGRATION emits tracker-migration-preview.v1 without mutating Tracker bytes.
package checklist

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

const trackerMigrationPreviewSchema = "tracker-migration-preview.v1"

const staleReasonStepRemoved = "step_removed_from_definition"

// StaleDisposition flags a Tracker row whose slug no longer exists in the definition.
type StaleDisposition struct {
	Slug        string `json:"slug"`
	Disposition string `json:"disposition"`
	Reason      string `json:"reason"`
}

// TrackerMigrationPreviewReport is the read-only slug diff between definition and Tracker.
type TrackerMigrationPreviewReport struct {
	SchemaVersion         string             `json:"schema_version"`
	DefinitionPath        string             `json:"definition_path"`
	TrackerPath           string             `json:"tracker_path"`
	ReadOnly              bool               `json:"read_only"`
	DefinitionSlugs       []string           `json:"definition_slugs"`
	TrackerSlugs          []string           `json:"tracker_slugs"`
	MissingInTracker      []string           `json:"missing_in_tracker"`
	ExtraInTracker        []string           `json:"extra_in_tracker"`
	StaleDispositions     []StaleDisposition `json:"stale_dispositions"`
	WouldMaterializeFresh bool               `json:"would_materialize_fresh"`
}

// DefinitionStepSlugs returns the slug inventory MaterializeAuthoritativeTracker would use.
func DefinitionStepSlugs(definitionPath string) ([]string, error) {
	data, err := os.ReadFile(definitionPath)
	if err != nil {
		return nil, fmt.Errorf("definition_not_readable: %w", err)
	}
	var doc yamlDoc
	if err := yaml.Unmarshal(data, &doc); err != nil {
		return nil, fmt.Errorf("invalid_definition: %w", err)
	}
	if doc.Steps == nil {
		return nil, fmt.Errorf("invalid_definition: missing steps in %s", definitionPath)
	}
	if err := validateStepsHaveSlugs(doc.Steps, definitionPath); err != nil {
		return nil, err
	}
	if err := validateDuplicateSlugs(doc.Steps, definitionPath); err != nil {
		return nil, fmt.Errorf("duplicate_slug: %w", err)
	}
	hasGateSub := false
	for _, sub := range doc.SubProcedures {
		if strings.TrimSpace(sub.Slug) == GateSubProcedureSlug {
			hasGateSub = true
			break
		}
	}
	if !hasGateSub {
		return nil, fmt.Errorf("missing_gate_sub_procedure: %q not found in sub_procedures", GateSubProcedureSlug)
	}

	slugs := make([]string, 0, len(doc.Steps)+1)
	seen := make(map[string]struct{}, len(doc.Steps)+1)
	for _, step := range doc.Steps {
		slug := strings.TrimSpace(step.Slug)
		if _, ok := seen[slug]; ok {
			return nil, fmt.Errorf("duplicate_slug: %q", slug)
		}
		seen[slug] = struct{}{}
		slugs = append(slugs, slug)
	}
	if _, ok := seen[GateSubProcedureSlug]; !ok {
		slugs = append(slugs, GateSubProcedureSlug)
	}
	return slugs, nil
}

// TrackerStepDispositions returns slug→disposition map from a Tracker document.
func TrackerStepDispositions(tracker map[string]interface{}) map[string]string {
	out := make(map[string]string)
	stepsRaw, ok := tracker["steps"].([]interface{})
	if !ok {
		return out
	}
	for _, item := range stepsRaw {
		row, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		slug, _ := row["slug"].(string)
		slug = strings.TrimSpace(slug)
		if slug == "" {
			continue
		}
		disp, _ := row["disposition"].(string)
		out[slug] = strings.TrimSpace(disp)
	}
	return out
}

// PreviewTrackerMigration compares definition and Tracker slug inventories without writing Tracker.
func PreviewTrackerMigration(definitionPath, trackerPath string) (TrackerMigrationPreviewReport, error) {
	report := TrackerMigrationPreviewReport{
		SchemaVersion:         trackerMigrationPreviewSchema,
		DefinitionPath:        filepath.Clean(definitionPath),
		TrackerPath:           filepath.Clean(trackerPath),
		ReadOnly:              true,
		WouldMaterializeFresh: false,
	}
	if err := RefuseDefinitionAsTracker(definitionPath, trackerPath); err != nil {
		return report, err
	}
	defSlugs, err := DefinitionStepSlugs(definitionPath)
	if err != nil {
		return report, err
	}
	report.DefinitionSlugs = append([]string(nil), defSlugs...)

	tracker, err := LoadTrackerYAML(trackerPath)
	if err != nil {
		return report, fmt.Errorf("tracker_not_readable: %w", err)
	}
	trackSlugs := TrackerStepSlugs(tracker)
	report.TrackerSlugs = append([]string(nil), trackSlugs...)

	defSet := slugSet(defSlugs)
	trackSet := slugSet(trackSlugs)
	report.MissingInTracker = sortedSetDiff(defSet, trackSet)
	report.ExtraInTracker = sortedSetDiff(trackSet, defSet)

	dispositions := TrackerStepDispositions(tracker)
	for _, slug := range report.ExtraInTracker {
		disp := dispositions[slug]
		if disp == "" || disp == "pending" {
			continue
		}
		report.StaleDispositions = append(report.StaleDispositions, StaleDisposition{
			Slug:        slug,
			Disposition: disp,
			Reason:      staleReasonStepRemoved,
		})
	}
	sort.Slice(report.StaleDispositions, func(i, j int) bool {
		return report.StaleDispositions[i].Slug < report.StaleDispositions[j].Slug
	})
	return report, nil
}

func slugSet(slugs []string) map[string]struct{} {
	out := make(map[string]struct{}, len(slugs))
	for _, slug := range slugs {
		out[slug] = struct{}{}
	}
	return out
}

func sortedSetDiff(from, subtract map[string]struct{}) []string {
	out := make([]string, 0)
	for slug := range from {
		if _, ok := subtract[slug]; ok {
			continue
		}
		out = append(out, slug)
	}
	sort.Strings(out)
	return out
}

File: mcp-server/packages/agentstream/src/tracker-migration-preview.ts
/**
 * [IMPL-TIED_UNIFIED_TOOLCHAIN] [ARCH-TIED_CHECKLIST_GATE_ENFORCEMENT] [REQ-TIED_CHECKLIST_GATE_ENFORCEMENT]
 * Read-only tracker migration preview (Go checklist/tracker_migration_preview.go parity).
 */
import fs from "node:fs";
import path from "node:path";
import yaml from "js-yaml";

import {
  GATE_SUB_PROCEDURE_SLUG,
  STALE_REASON_STEP_REMOVED,
  TRACKER_MIGRATION_PREVIEW_SCHEMA,
} from "./checklist-constants.js";
import {
  loadTrackerYaml,
  trackerStepDispositions,
  trackerStepSlugs,
} from "./tracker-yaml.js";

type ChecklistStep = { slug?: string };
type ChecklistSub = { slug?: string };
type ChecklistDoc = {
  steps?: ChecklistStep[];
  sub_procedures?: ChecklistSub[];
};

export type StaleDisposition = {
  slug: string;
  disposition: string;
  reason: string;
};

export type TrackerMigrationPreviewReport = {
  schema_version: string;
  definition_path: string;
  tracker_path: string;
  read_only: boolean;
  definition_slugs: string[];
  tracker_slugs: string[];
  missing_in_tracker: string[];
  extra_in_tracker: string[];
  stale_dispositions: StaleDisposition[] | null;
  would_materialize_fresh: boolean;
};

export function refuseDefinitionAsTracker(
  definitionPath: string,
  trackerPath: string,
): void {
  const def = path.resolve(path.normalize(definitionPath));
  const track = path.resolve(path.normalize(trackerPath));
  if (def === track) {
    throw new Error(
      `tracker path must not equal checklist definition path: ${def}`,
    );
  }
}

function validateStepsHaveSlugs(steps: ChecklistStep[], filePath: string): void {
  for (let i = 0; i < steps.length; i++) {
    if (String(steps[i]?.slug ?? "").trim() === "") {
      throw new Error(
        `checklist step at "steps" index ${i} missing required "slug" in ${filePath}`,
      );
    }
  }
}

function validateDuplicateSlugs(steps: ChecklistStep[], filePath: string): void {
  const seen = new Map<string, number>();
  for (let i = 0; i < steps.length; i++) {
    const slug = String(steps[i]?.slug ?? "").trim();
    const prev = seen.get(slug);
    if (prev !== undefined) {
      throw new Error(
        `checklist duplicate slug ${JSON.stringify(slug)} at "steps" indices ${prev} and ${i} in ${filePath}`,
      );
    }
    seen.set(slug, i);
  }
}

export function definitionStepSlugs(definitionPath: string): string[] {
  let data: string;
  try {
    data = fs.readFileSync(definitionPath, "utf8");
  } catch (err) {
    throw new Error(`definition_not_readable: ${String(err)}`);
  }
  let doc: ChecklistDoc;
  try {
    doc = yaml.load(data) as ChecklistDoc;
  } catch (err) {
    throw new Error(`invalid_definition: ${String(err)}`);
  }
  if (doc.steps === undefined || doc.steps === null) {
    throw new Error(`invalid_definition: missing steps in ${definitionPath}`);
  }
  validateStepsHaveSlugs(doc.steps, definitionPath);
  try {
    validateDuplicateSlugs(doc.steps, definitionPath);
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    throw new Error(`duplicate_slug: ${msg}`);
  }

  let hasGateSub = false;
  for (const sub of doc.sub_procedures ?? []) {
    if (String(sub.slug ?? "").trim() === GATE_SUB_PROCEDURE_SLUG) {
      hasGateSub = true;
      break;
    }
  }
  if (!hasGateSub) {
    throw new Error(
      `missing_gate_sub_procedure: ${JSON.stringify(GATE_SUB_PROCEDURE_SLUG)} not found in sub_procedures`,
    );
  }

  const slugs: string[] = [];
  const seen = new Set<string>();
  for (const step of doc.steps) {
    const slug = String(step.slug ?? "").trim();
    if (seen.has(slug)) {
      throw new Error(`duplicate_slug: ${JSON.stringify(slug)}`);
    }
    seen.add(slug);
    slugs.push(slug);
  }
  if (!seen.has(GATE_SUB_PROCEDURE_SLUG)) {
    slugs.push(GATE_SUB_PROCEDURE_SLUG);
  }
  return slugs;
}

function slugSet(slugs: string[]): Set<string> {
  return new Set(slugs);
}

function sortedSetDiff(from: Set<string>, subtract: Set<string>): string[] {
  const out: string[] = [];
  for (const slug of from) {
    if (subtract.has(slug)) {
      continue;
    }
    out.push(slug);
  }
  out.sort();
  return out;
}

export function previewTrackerMigration(
  definitionPath: string,
  trackerPath: string,
): TrackerMigrationPreviewReport {
  const report: TrackerMigrationPreviewReport = {
    schema_version: TRACKER_MIGRATION_PREVIEW_SCHEMA,
    definition_path: path.normalize(definitionPath),
    tracker_path: path.normalize(trackerPath),
    read_only: true,
    definition_slugs: [],
    tracker_slugs: [],
    missing_in_tracker: [],
    extra_in_tracker: [],
    stale_dispositions: null,
    would_materialize_fresh: false,
  };

  refuseDefinitionAsTracker(definitionPath, trackerPath);

  const defSlugs = definitionStepSlugs(definitionPath);
  report.definition_slugs = [...defSlugs];

  let tracker;
  try {
    tracker = loadTrackerYaml(trackerPath);
  } catch (err) {
    throw new Error(`tracker_not_readable: ${String(err)}`);
  }

  const trackSlugs = trackerStepSlugs(tracker);
  report.tracker_slugs = [...trackSlugs];

  const defSet = slugSet(defSlugs);
  const trackSet = slugSet(trackSlugs);
  report.missing_in_tracker = sortedSetDiff(defSet, trackSet);
  report.extra_in_tracker = sortedSetDiff(trackSet, defSet);

  const dispositions = trackerStepDispositions(tracker);
  const stale: StaleDisposition[] = [];
  for (const slug of report.extra_in_tracker) {
    const disp = dispositions[slug] ?? "";
    if (disp === "" || disp === "pending") {
      continue;
    }
    stale.push({
      slug,
      disposition: disp,
      reason: STALE_REASON_STEP_REMOVED,
    });
  }
  stale.sort((a, b) => a.slug.localeCompare(b.slug));
  report.stale_dispositions = stale.length > 0 ? stale : null;

  return report;
}

export function encodePreviewReport(
  report: TrackerMigrationPreviewReport,
): string {
  return `${JSON.stringify(report, null, 2)}\n`;
}
